pkg/profiler/profiler.go

```diff
@@ -372,69 +374,52 @@ func (p *CgroupProfiler) Run(ctx context.Context) error {
 	}
 }
 
-func (p *CgroupProfiler) profileLoop(ctx context.Context, captureTime time.Time) error {
-	prof := &profile.Profile{
-		SampleType: []*profile.ValueType{{
-			Type: "samples",
-			Unit: "count",
-		}},
-		TimeNanos:     captureTime.UnixNano(),
-		DurationNanos: int64(p.profilingDuration),
-
-		// We sample at 100Hz, which is every 10 Million nanoseconds.
-		PeriodType: &profile.ValueType{
-			Type: "cpu",
-			Unit: "nanoseconds",
-		},
-		Period: 10000000,
-	}
-
-	mapping := maps.NewMapping(p.pidMappingFileCache)
-	kernelMapping := &profile.Mapping{
-		// TODO(kakkoyun): Check if this conflicts with https://github.com/google/pprof/pull/675/files
-		File: "[kernel.kallsyms]",
-	}
-	kernelFunctions := map[uint64]*profile.Function{}
-	userFunctions := map[[2]uint64]*profile.Function{}
+func (p *CgroupProfiler) profileLoop(ctx context.Context, captureTime time.Time) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			level.Error(p.logger).Log("msg", "panic in profile loop", "err", r)
+			debug.PrintStack()
+			err = fmt.Errorf("panic in profile loop")
+		}
+	}()
 
-	// 2 uint64 1 for PID and 1 for Addr
-	locations := []*profile.Location{}
-	kernelLocations := []*profile.Location{}
-	kernelAddresses := map[uint64]struct{}{}
-	locationIndices := map[[2]uint64]int{}
-	samples := map[[doubleStackDepth]uint64]*profile.Sample{}
+	var (
+		mappings      = maps.NewMapping(p.pidMappingFileCache)
+		kernelMapping = &profile.Mapping{
+			// TODO(kakkoyun): Check if this conflicts with https://github.com/google/pprof/pull/675/files
+			File: "[kernel.kallsyms]",
+		}
 
-	it := p.bpfMaps.counts.Iterator()
+		samples         = map[[doubleStackDepth]uint64]*profile.Sample{}
+		locations       = []*profile.Location{}
+		kernelLocations = []*profile.Location{}
+		userLocations   = map[uint32][]*profile.Location{} // PID -> []*profile.Location
+		locationIndices = map[[2]uint64]int{}              // [PID, Address] -> index in locations
+	)
 
 	// TODO(brancz): Use libbpf batch functions.
+	it := p.bpfMaps.counts.Iterator()
 	for it.Next() {
 		// This byte slice is only valid for this iteration, so it must be
-		// copied if we want to do anything with it outside of this loop.
+		// copied if we want to do anything with it outside this loop.
 		keyBytes := it.Key()
-		r := bytes.NewBuffer(keyBytes)
 
-		pidBytes := make([]byte, 4)
-		if _, err := io.ReadFull(r, pidBytes); err != nil {
-			return fmt.Errorf("read pid bytes: %w", err)
-		}
-		pid := p.byteOrder.Uint32(pidBytes)
-		if pid == 0 {
-			level.Debug(p.logger).Log("msg", "missing pid")
-			p.metrics.missingPIDs.Inc()
-			continue
+		var key stackCountKey
+		if err := binary.Read(bytes.NewBuffer(keyBytes), p.byteOrder, &key); err != nil {
```

Comment:
Are you 100% sure this works as expected? I remember trying something like this but running into problems because of automatic byte alignment that the Go compiler applies to the `stackCountKey` struct.

Reply:
This just work because the struct we have is perfectly aligned! So I will add a notice in case we need to change it we should think about alignment.

https://dave.cheney.net/2015/10/09/padding-is-hard